simulation/utils/generate_map.py

```diff
@@ -0,0 +1,172 @@
+        left_limit = i - ran if i >= ran else 0
+        right_limit = i + ran if i + ran < size else size - 2
+        for j in range(left_limit + 1, right_limit):
+            j = size - 2 if j >= size - 1 else j
+            if is_crossing(X, Y, i, j):
+                del X[i + 1]
+                del Y[i + 1]
+                x, y = next_point(X[size - 2], Y[size - 2], step, math.pi * 3 / 4, y_res)
+                X.append(x)
+                Y.append(y)
+        i += 1
+    return X, Y
+
+
+def prepare_map(x_res, y_res, step, angle_range, starting_y):
+    """Function that prepares real map - interpolates random points generated by function next_point
+        returns interpolated and smoothed map in X and Y arrays
+    """
+    X = []
+    Y = []
+    x = 0
+    y = random.randrange(y_res / 4, y_res * 3 / 5) if not starting_y else starting_y
+    X.append(x)
+    Y.append(y)
+
+    while x < x_res:
+        x, y = next_point(x, y, step, angle_range, y_res)
+        X.append(x)
+        Y.append(y)
+
+    X, Y = exterminate_loops(X, Y, step, y_res)
+    X, Y = exterminate_loops(X, Y, step, y_res)
+    X, Y = exterminate_loops(X, Y, step, y_res)
+
+    tck, u = splprep([X, Y], s=0)
+    x_array = np.linspace(0, 1, x_res)
+    x_array, y_array = splev(x_array, tck, der=0)
+
+    return x_array, y_array
+
+
+def get_road(diff_level, starting_y, resolution):
+    """Function that prepares parameters according to difficulty level of map
+        returns X and Y ready to go
+    """
+    x_res = resolution[0]
+    y_res = resolution[1]
+    x_array, y_array = None, None
+
+    if diff_level == Difficulty.PATHETIC:
+        x = np.array([0, x_res])
+        x_array = np.linspace(x.min(), x.max(), x_res)
+        y_start = y_res / 2 if not starting_y else starting_y
+        y_array = np.array([y_start for _ in range(len(x_array))])
+
+    elif diff_level == Difficulty.EASY:
+        step = x_res / 10
+        angle_range = math.pi / 4
+        x_array, y_array = prepare_map(x_res, y_res, step, angle_range, starting_y)
+
+    elif diff_level == Difficulty.MEDIUM:
+        step = x_res / 20
+        angle_range = math.pi * 3 / 4
+        x_array, y_array = prepare_map(x_res, y_res, step, angle_range, starting_y)
+
+    elif diff_level == Difficulty.HARD:
+        step = x_res / 100
+        angle_range = 15 * math.pi / 18
+        x_array, y_array = prepare_map(x_res, y_res, step, angle_range, starting_y)
+
+    elif diff_level == Difficulty.REALLY_HARD:
+        step = x_res / 120
+        angle_range = 16 * math.pi / 18
+        x_array, y_array = prepare_map(x_res, y_res, step, angle_range, starting_y)
+
+    elif diff_level == Difficulty.WTF:
+        step = x_res / 150
+        angle_range = 16 * math.pi / 18
+        x_array, y_array = prepare_map(x_res, y_res, step, angle_range, starting_y)
+
+    return x_array, y_array
+
+
+def generate_map(seed=None, diff_level=Difficulty.PATHETIC, x_offset=0, starting_y=None, resolution=(1280, 720)):
```

Comment:
Skoro diff_level i resolution są najbardziej doprecyzowane to przesunąłbym je w prawo na liście argumentów. O ile dobrze rozumiem bez starting_y ciężko będzie zacząć działanie, a w diff_level możemy używać łatwo wartości domyślnej. No i jeśli rzeczywiście brak tej wartości wyrzuci błąd albo sprawi że nie będzie to działać dobrze to wywalić wartość domyślną